driver-core/src/main/com/mongodb/connection/Server.java

```diff
@@ -54,5 +54,5 @@
      *
      * @param callback the callback to execute when the connection is available or an error occurs
      */
-    void getConnectionAsync(SingleResultCallback<Connection> callback);
+    void getConnectionAsync(SingleResultCallback<AsyncConnection> callback);
```

Comment:
I think this should be called getAsyncConnection or getAsyncConnectionAsync.

I was surprised in AsyncSingleConnectionBinding that getConnectionAsync returned an AsyncConnection. Simply because I wanted to get a connection in an async manner doesn't necessarily mean I want an async connection. 

I'd even go so far as to say you should support getting an AsyncConnection synchronously. Case in point is that in AsyncSingleConnectionBinding, you are getting read and write connections in the ctor (which can't be async) by using countdown latches. It would be much simpler to just use a sync API here.